[DOC] KafkaConnector resources (#2338)
strimzi/strimzi-kafka-operator: documentation/modules/con-creating-managing-connectors.adoc

// Module included in the following assemblies:
//
// assembly-kafka-connect.adoc

[id='con-creating-managing-connectors-{context}']

= Creating and managing connectors

> Once you have created a container image for your connector plug-in, you need to create a connector instance in your Kafka Connect cluster. You can then configure, monitor, and manage a running connector instance. For example, you can:

Reviewer: Consider 'When' or 'After' rather than 'Once', which can mean 'one time'.

* Check the status of a connector instance
* Increase or decrease the number of tasks
* Restart failed tasks
* Pause a connector instance
* Delete a connector instance

{ProductName} provides two different APIs for creating and managing connectors:

> * `KafkaConnector` resources (referred to as `KafkaConnectors`)

Reviewer 2: Do we _really_ need the "referred to as"? We're just pluralising, after all?

Author: Might be helpful for non-native English speakers.

* Kafka Connect REST API

NOTE: Currently, `KafkaConnectors` do not support restarting failed tasks and pausing and resuming connectors. You need to use the Kafka Connect REST API to perform these tasks.

.`KafkaConnector` resources

> `KafkaConnectors` allow you to create and manage source and sink connector instances for Kafka Connect in a Kubernetes-native way, so an HTTP client such as cURL is not required. Like other Kafka resources, you declare a connector’s desired state in a `KafkaConnector` YAML file that is deployed to your Kubernetes cluster to create the connector instance. 

Reviewer 2: ```suggestion
`KafkaConnectors` allow you to create and manage connector instances for Kafka Connect in a Kubernetes-native way, so an HTTP client such as cURL is not required.
Like other Kafka resources, you declare a connector’s desired state in a `KafkaConnector` YAML file that is deployed to your Kubernetes cluster to create the connector instance. 
```

> You manage a running connector instance by updating its corresponding `KafkaConnector`, and then applying the updates (for example, using `kubectl apply`). You remove a connector by deleting its corresponding `KafkaConnector`.

Reviewer: I wonder if we need the example in brackets. If we do, maybe a whole example?

// Module included in the following assemblies:
//
// assembly-kafka-connect.adoc

[id='con-creating-managing-connectors-{context}']

= Creating and managing connectors

Once you have created a container image for your connector plug-in, you need to create a connector instance in your Kafka Connect cluster. You can then configure, monitor, and manage a running connector instance. For example, you can:

* Check the status of a connector instance
* Increase or decrease the number of tasks
* Restart failed tasks
* Pause a connector instance
* Delete a connector instance

{ProductName} provides two different APIs for creating and managing connectors:

* `KafkaConnector` resources (referred to as `KafkaConnectors`)
* Kafka Connect REST API

> For security reasons, we recommend that you use `KafkaConnectors` rather than the Kafka Connect REST API. `KafkaConnectors` use Kubernetes RBAC controls to secure the Kafka Connect REST API and lock down its resources.

Reviewer 2: This isn't true and the messaging here is a little bit difficult. We want users to start using `KafkaConnectors` because we think that eventually we not expose the REST API at all (or at least not by default). But right now it doesn't support pause/resume or task restart, so it's not really feasible to lock it down.

Using NetworkPolicies they will be able to limit where requests to the Kafka Connect REST API come from. And using RBAC they will be able to control who has access the connectors. So the access control story is _potentially_ better using `KafkaConnector` than without. But it's not currently better by default.

Author: Thanks for the explanation. It's tricky alright. Rather than mention security at all, I suggest that we just have a single note.

NOTE: Currently, `KafkaConnectors` do not support restarting failed tasks and pausing and resuming connectors. You need to use the Kafka Connect REST API to perform these tasks.

NOTE: You need to use the Kafka Connect REST API for restarting failed tasks and pausing and resuming connectors. Future {ProductName} releases might add this functionality to `KafkaConnectors`.

.`KafkaConnector` resources

> `KafkaConnectors` allow you to create and manage source and sink connector instances for Kafka Connect in a Kubernetes-native way, so an HTTP client such as cURL is not required. Like other Kafka resources, you declare a connector’s desired state in a `KafkaConnector` YAML file that is deployed to your Kubernetes cluster to create a corresponding connector instance. 

Reviewer 2: This is not incorrect, but (as a general point) I don't especially like the terminology of "deploying a YAML file to Kubernetes". It's only a file if you're using a file to represent the resource locally. There are other ways of interacting with the Kubernetes API server which aren't so much about files (e.g. `kubectl edit`, various consoles etc).

> You manage a running connector instance by updating its corresponding `KafkaConnector`, and then applying the updates using `kubectl apply`. You remove a connector by deleting its corresponding `KafkaConnector`.

Reviewer 2: ```suggestion
You manage a running connector instance by updating its corresponding `KafkaConnector`, and then applying the updates, using `kubectl apply` for example. You remove a connector by deleting its corresponding `KafkaConnector`.
```

To ensure compatibility with earlier versions of {ProductName}, `KafkaConnectors` are disabled by default. To enable them for a Kafka Connect cluster, you must configure the `KafkaConnect` custom resource to use `KafkaConnectors` for the creation and management of connectors. For instructions, see xref:proc-enabling-kafkaconnectors-deployment-configuration-kafka-connect[].

When `KafkaConnectors` are enabled, the Cluster Operator begins to watch for them. Like other Kafka resources, the Cluster Operator updates the configurations of running connector instances to match the configurations defined in their `KafkaConnectors`.

.Availability of the Kafka Connect REST API

The Kafka Connect REST API is available on port 8083 as the `<connect-cluster-name>-connect-api` service.

> If `KafkaConnectors` are enabled, changes made using the Kafka Connect REST API are reverted by the Cluster Operator. 

Reviewer 2: ```suggestion
If `KafkaConnectors` are enabled, manual changes made directly using the Kafka Connect REST API are reverted by the Cluster Operator. 
```

// Module included in the following assemblies:
//
// assembly-kafka-connect.adoc

[id='con-creating-managing-connectors-{context}']

= Creating and managing connectors

When you have created a container image for your connector plug-in, you need to create a connector instance in your Kafka Connect cluster. You can then configure, monitor, and manage a running connector instance. For example, you can:

* Check the status of a connector instance
* Increase or decrease the number of tasks
* Restart failed tasks
* Pause a connector instance
* Delete a connector instance

{ProductName} provides two different APIs for creating and managing connectors:

* `KafkaConnector` resources (referred to as `KafkaConnectors`)
* Kafka Connect REST API

> NOTE: Currently, `KafkaConnectors` do not support restarting failed tasks or pausing and resuming connectors. You need to use the Kafka Connect REST API to perform those tasks.

Reviewer 2: Support for pause/resume has been merged...

```suggestion
NOTE: Currently, `KafkaConnectors` do not support restarting failed tasks. You need to use the Kafka Connect REST API to to this.
```

Author: OK, that's updated.
